middleware/proxy: data race

There is still a data race here. Inclined to revert stuff to get this back to normal :(
(/cc @mfischer-zd this is still racy, I thought I had fixed it lately, but no dice)

~~~
==================
WARNING: DATA RACE
Write at 0x00c420231cec by goroutine 57:
  github.com/coredns/coredns/middleware/proxy.(*staticUpstream).healthCheck()
      /home/travis/gopath/src/github.com/coredns/coredns/middleware/proxy/upstream.go:262 +0x697
  github.com/coredns/coredns/middleware/proxy.(*staticUpstream).HealthCheckWorker()
      /home/travis/gopath/src/github.com/coredns/coredns/middleware/proxy/upstream.go:273 +0x84
Previous read at 0x00c420231cec by goroutine 124:
  github.com/coredns/coredns/middleware/proxy.NewStaticUpstreams.func1.1()
      /home/travis/gopath/src/github.com/coredns/coredns/middleware/proxy/upstream.go:91 +0x65
  github.com/coredns/coredns/middleware/proxy.(*UpstreamHost).Down()
      /home/travis/gopath/src/github.com/coredns/coredns/middleware/proxy/proxy.go:73 +0xeb
  github.com/coredns/coredns/middleware/proxy.(*staticUpstream).Select()
      /home/travis/gopath/src/github.com/coredns/coredns/middleware/proxy/upstream.go:289 +0x3e7
  github.com/coredns/coredns/middleware/proxy.Proxy.ServeDNS()
      /home/travis/gopath/src/github.com/coredns/coredns/middleware/proxy/proxy.go:97 +0x254
  github.com/coredns/coredns/middleware/proxy.(*Proxy).ServeDNS()
      <autogenerated>:12 +0xec
  github.com/coredns/coredns/core/dnsserver.(*Server).ServeDNS()
      /home/travis/gopath/src/github.com/coredns/coredns/core/dnsserver/server.go:190 +0xab9
  github.com/coredns/coredns/core/dnsserver.(*Server).ServePacket.func1()
      /home/travis/gopath/src/github.com/coredns/coredns/core/dnsserver/server.go:88 +0xb9
  github.com/miekg/dns.HandlerFunc.ServeDNS()
      /home/travis/gopath/src/github.com/miekg/dns/server.go:84 +0x51
  github.com/miekg/dns.(*Server).serve()
      /home/travis/gopath/src/github.com/miekg/dns/server.go:579 +0x80c
Goroutine 57 (running) created at:
  github.com/coredns/coredns/middleware/proxy.NewStaticUpstreams()
      /home/travis/gopath/src/github.com/coredns/coredns/middleware/proxy/upstream.go:109 +0xadd
  github.com/coredns/coredns/middleware/proxy.setup()
      /home/travis/gopath/src/github.com/coredns/coredns/middleware/proxy/setup.go:18 +0x72
  github.com/mholt/caddy.executeDirectives()
      /home/travis/gopath/src/github.com/mholt/caddy/caddy.go:583 +0x794
  github.com/mholt/caddy.ValidateAndExecuteDirectives()
      /home/travis/gopath/src/github.com/mholt/caddy/caddy.go:533 +0x4e2
  github.com/mholt/caddy.startWithListenerFds()
      /home/travis/gopath/src/github.com/mholt/caddy/caddy.go:437 +0x7b
  github.com/mholt/caddy.Start()
      /home/travis/gopath/src/github.com/mholt/caddy/caddy.go:429 +0x14d
  github.com/coredns/coredns/test.CoreDNSServer()
      github.com/coredns/coredns/test/_test/_obj_test/server.go:27 +0x129
  github.com/coredns/coredns/test.TestProxyWithHTTPCheckOK()
      /home/travis/gopath/src/github.com/coredns/coredns/test/proxy_http_health_test.go:73 +0x57f
  testing.tRunner()
      /home/travis/.gimme/versions/go1.8.linux.amd64/src/testing/testing.go:657 +0x107
Goroutine 124 (running) created at:
  github.com/miekg/dns.(*Server).serveUDP()
      /home/travis/gopath/src/github.com/miekg/dns/server.go:533 +0x38c
  github.com/miekg/dns.(*Server).ActivateAndServe()
      /home/travis/gopath/src/github.com/miekg/dns/server.go:400 +0x420
  github.com/coredns/coredns/core/dnsserver.(*Server).ServePacket()
      /home/travis/gopath/src/github.com/coredns/coredns/core/dnsserver/server.go:92 +0x20d
  github.com/mholt/caddy.startServers.func1()
      /home/travis/gopath/src/github.com/mholt/caddy/caddy.go:672 +0xfc
==================
--- FAIL: TestProxyWithHTTPCheckOK (0.01s)
	testing.go:610: race detected during execution of test
~~~

--- a/middleware/proxy/google.go
+++ b/middleware/proxy/google.go
@@ -218,11 +218,11 @@ func newUpstream(hosts []string, old *staticUpstream) Upstream {
 			Conns:       0,
 			Fails:       0,
 			FailTimeout: upstream.FailTimeout,
-			Unhealthy:   newBool(),
+			Unhealthy:   false,
 
 			CheckDown: func(upstream *staticUpstream) UpstreamHostDownFunc {
 				return func(uh *UpstreamHost) bool {
-					if *uh.Unhealthy {
+					if uh.Unhealthy {
 						return true
 					}
 

--- a/middleware/proxy/lookup.go
+++ b/middleware/proxy/lookup.go
@@ -38,10 +38,10 @@ func NewLookupWithOption(hosts []string, opts Options) Proxy {
 			Fails:       0,
 			FailTimeout: upstream.FailTimeout,
 
-			Unhealthy: newBool(),
+			Unhealthy: false,
 			CheckDown: func(upstream *staticUpstream) UpstreamHostDownFunc {
 				return func(uh *UpstreamHost) bool {
-					if *uh.Unhealthy {
+					if uh.Unhealthy {
 						return true
 					}
 					fails := atomic.LoadInt32(&uh.Fails)

--- a/middleware/proxy/policy_test.go
+++ b/middleware/proxy/policy_test.go
@@ -28,16 +28,13 @@ func (r *customPolicy) Select(pool HostPool) *UpstreamHost {
 func testPool() HostPool {
 	pool := []*UpstreamHost{
 		{
-			Name:      workableServer.URL, // this should resolve (healthcheck test)
-			Unhealthy: newBool(),
+			Name: workableServer.URL, // this should resolve (healthcheck test)
 		},
 		{
-			Name:      "http://shouldnot.resolve", // this shouldn't
-			Unhealthy: newBool(),
+			Name: "http://shouldnot.resolve", // this shouldn't
 		},
 		{
-			Name:      "http://C",
-			Unhealthy: newBool(),
+			Name: "http://C",
 		},
 	}
 	return HostPool(pool)
@@ -57,7 +54,7 @@ func TestRoundRobinPolicy(t *testing.T) {
 		t.Error("Expected second round robin host to be third host in the pool.")
 	}
 	// mark host as down
-	*pool[0].Unhealthy = true
+	pool[0].Unhealthy = true
 	h = rrPolicy.Select(pool)
 	if h != pool[1] {
 		t.Error("Expected third round robin host to be first host in the pool.")

--- a/middleware/proxy/proxy.go
+++ b/middleware/proxy/proxy.go
@@ -3,6 +3,7 @@ package proxy
 
 import (
 	"errors"
+	"sync"
 	"sync/atomic"
 	"time"
 
@@ -56,9 +57,10 @@ type UpstreamHost struct {
 	Name              string // IP address (and port) of this upstream host
 	Fails             int32
 	FailTimeout       time.Duration
-	Unhealthy         *bool
+	Unhealthy         bool
 	CheckDown         UpstreamHostDownFunc
 	WithoutPathPrefix string
+	checkMu           sync.Mutex
 }
 
 // Down checks whether the upstream host is down or not.
@@ -68,7 +70,7 @@ func (uh *UpstreamHost) Down() bool {
 	if uh.CheckDown == nil {
 		// Default settings
 		fails := atomic.LoadInt32(&uh.Fails)
-		return *uh.Unhealthy || fails > 0
+		return uh.Unhealthy || fails > 0
 	}
 	return uh.CheckDown(uh)
 }

--- a/middleware/proxy/upstream.go
+++ b/middleware/proxy/upstream.go
@@ -84,11 +84,11 @@ func NewStaticUpstreams(c *caddyfile.Dispenser) ([]Upstream, error) {
 				Conns:       0,
 				Fails:       0,
 				FailTimeout: upstream.FailTimeout,
-				Unhealthy:   newBool(),
+				Unhealthy:   false,
 
 				CheckDown: func(upstream *staticUpstream) UpstreamHostDownFunc {
 					return func(uh *UpstreamHost) bool {
-						if *uh.Unhealthy {
+						if uh.Unhealthy {
 							return true
 						}
 
@@ -251,19 +251,22 @@ func (u *staticUpstream) healthCheck() {
 
 		hostURL := "http://" + net.JoinHostPort(checkHostName, checkPort) + u.HealthCheck.Path
 
+		host.checkMu.Lock()
+		defer host.checkMu.Unlock()
+
 		if r, err := http.Get(hostURL); err == nil {
 			io.Copy(ioutil.Discard, r.Body)
 			r.Body.Close()
 			if r.StatusCode < 200 || r.StatusCode >= 400 {
 				log.Printf("[WARNING] Health check URL %s returned HTTP code %d\n",
 					hostURL, r.StatusCode)
-				*host.Unhealthy = true
+				host.Unhealthy = true
 			} else {
-				*host.Unhealthy = false
+				host.Unhealthy = false
 			}
 		} else {
 			log.Printf("[WARNING] Health check probe failed: %v\n", err)
-			*host.Unhealthy = true
+			host.Unhealthy = true
 		}
 	}
 }
@@ -338,9 +341,3 @@ func (u *staticUpstream) IsAllowedDomain(name string) bool {
 }
 
 func (u *staticUpstream) Exchanger() Exchanger { return u.ex }
-
-func newBool() *bool {
-	b := new(bool)
-	*b = false
-	return b
-}

--- a/middleware/proxy/upstream.go
+++ b/middleware/proxy/upstream.go
@@ -88,6 +88,8 @@ func NewStaticUpstreams(c *caddyfile.Dispenser) ([]Upstream, error) {
 
 				CheckDown: func(upstream *staticUpstream) UpstreamHostDownFunc {
 					return func(uh *UpstreamHost) bool {
+						uh.checkMu.Lock()
+						defer uh.checkMu.Unlock()
 						if uh.Unhealthy {
 							return true
 						}

--- a/middleware/proxy/upstream_test.go
+++ b/middleware/proxy/upstream_test.go
@@ -42,13 +42,13 @@ func TestSelect(t *testing.T) {
 		FailTimeout: 10 * time.Second,
 		MaxFails:    1,
 	}
-	*upstream.Hosts[0].Unhealthy = true
-	*upstream.Hosts[1].Unhealthy = true
-	*upstream.Hosts[2].Unhealthy = true
+	upstream.Hosts[0].Unhealthy = true
+	upstream.Hosts[1].Unhealthy = true
+	upstream.Hosts[2].Unhealthy = true
 	if h := upstream.Select(); h != nil {
 		t.Error("Expected select to return nil as all host are down")
 	}
-	*upstream.Hosts[2].Unhealthy = false
+	upstream.Hosts[2].Unhealthy = false
 	if h := upstream.Select(); h == nil {
 		t.Error("Expected select to not return nil")
 	}

